google/cloud/storage/tests/object_integration_test.cc

```diff
@@ -1146,6 +1143,116 @@ TEST_F(ObjectIntegrationTest, RewriteLarge) {
   client.DeleteObject(bucket_name, source_name);
 }
 
+/// @test Verify that MD5 hashes are computed by default.
+TEST_F(ObjectIntegrationTest, DefaultMD5HashXML) {
+  auto backend = std::make_shared<CaptureSendHeaderBackend>();
+
+  Client client(ClientOptions()
+                    .set_enable_raw_client_tracing(true)
+                    .set_enable_http_tracing(true));
+  auto bucket_name = ObjectTestEnvironment::bucket_name();
+  auto object_name = MakeRandomObjectName();
+
+  auto id = LogSink::Instance().AddBackend(backend);
+  ObjectMetadata insert_meta = client.InsertObject(
+      bucket_name, object_name, LoremIpsum(), IfGenerationMatch(0), Fields(""));
+  LogSink::Instance().RemoveBackend(id);
+
+  long count =
+      std::count_if(backend->log_lines.begin(), backend->log_lines.end(),
+                    [](std::string const& line) {
+                      return line.rfind("x-goog-hash: md5=", 0) == 0;
+                    });
+  EXPECT_EQ(1, count);
+
+  client.DeleteObject(bucket_name, object_name);
+}
+
+/// @test Verify that MD5 hashes are computed by default.
+TEST_F(ObjectIntegrationTest, DefaultMD5HashJSON) {
+  auto backend = std::make_shared<CaptureSendHeaderBackend>();
+
+  Client client(ClientOptions()
+                    .set_enable_raw_client_tracing(true)
+                    .set_enable_http_tracing(true));
+  auto bucket_name = ObjectTestEnvironment::bucket_name();
+  auto object_name = MakeRandomObjectName();
+
+  auto id = LogSink::Instance().AddBackend(backend);
+  ObjectMetadata insert_meta = client.InsertObject(
+      bucket_name, object_name, LoremIpsum(), IfGenerationMatch(0));
+  LogSink::Instance().RemoveBackend(id);
+
+  long count = std::count_if(
+      backend->log_lines.begin(), backend->log_lines.end(),
+      [](std::string const& line) {
+        // This is a big indirect, we detect if the upload changed to
+        // multipart/related, and if so, we assume the hash value is being used.
+        // Unfortunately I (@coryan) cannot think of a way to examine the upload
+        // contents.
+        return line.rfind("content-type: multipart/related; boundary=", 0) == 0;
```

Comment:
Maybe when we add the MD5 to the request body, we could also print out the computed MD5 with `GCP_LOG`, then check those lines for that checksum?  This still wouldn't ensure that the MD5 is actually *in* the request, but would at least tell us that we computed the correct MD5 and got to the logging statement directly next to the statement that actually puts it in the request body. /shrug

Reply:
>  but would at least tell us that we computed the correct MD5

We know we have the correct MD5 because GCS rejects the request when we send the wrong one: we trust GCS to do its work -- at least I do after I tested with an intentionally broken MD5 :grin:

> This still wouldn't ensure that the MD5 is actually *in* the request, 

What if we captured that sort of information in the testbench, and returned it as `metadata.x_testbench_*` fields in object metadata.  We can then examine those extra attributes in the test?

Reply:
That would be fine if we only checked for it when running against the emulator (i.e. this would fail when running tests against prod).

Reply:
Done. And rebased to resolve conflicts.